### PR TITLE
GEOS-7392 SecuredLayerGroup bug

### PR DESCRIPTION
**Pull request by NielsCharlier**
https://osgeo-org.atlassian.net/browse/GEOS-7392

ModificationProxyCloner change was in the end not necessary to fix the bug, but the change still makes sense so I am leaving it in.

### CODE REVIEW

**Comment by NielsCharlier** [comment]
the second change to the ModificationProxyCloner is necessary.

**Comment by aaime** [comment]
The Travis failures appear to be legit and caused by this change (at the very least, they do involve layer groups):

````
Failed tests:   testRecordSortedAndPaged(org.geoserver.csw.store.internal.LayerGroupTest): Expecting to find matches for Xpath //csw:SummaryRecord[dc:title='Buildings and Bridges']
  testRecords(org.geoserver.csw.store.internal.LayerGroupTest): Expecting to find matches for Xpath //csw:Record[dc:title='Buildings and Bridges']

Tests in error: 
  testRecordById(org.geoserver.csw.store.internal.LayerGroupTest): org.geoserver.security.decorators.SecuredLayerGroupInfo cannot be cast to org.geoserver.catalog.impl.LayerGroupInfoImpl
````

**Comment by NielsCharlier** [comment]
fixed

File: src/extension/csw/core/src/test/java/org/geoserver/csw/store/internal/LayerGroupTest.java [hunk 1]
```diff
@@ -5,8 +5,8 @@
```
Comment (aaime) [inline src/extension/csw/core/src/test/java/org/geoserver/csw/store/internal/LayerGroupTest.java]: Lack copyright header

File: src/extension/csw/core/src/test/java/org/geoserver/csw/store/internal/LayerGroupTest.java [hunk 2]
```diff
@@ -61,7 +62,7 @@ public void testRecords() throws Exception {
     
     @Test
     public void testRecordById() throws Exception {
-        String request = "csw?service=CSW&version=2.0.2&request=GetRecordById&typeNames=csw:Record&id=" + ID_FORESTSANDSTREAMS;
+        String request = "csw?service=CSW&version=2.0.2&request=GetRecordById&typeNames=csw:Record&id=" + id_forestsandstreams;
         Document d = getAsDOM(request);
         //print(d);
         checkValidationErrors(d);
```
Comment (aaime) [inline src/extension/csw/core/src/test/java/org/geoserver/csw/store/internal/LayerGroupTest.java]: testRecordSortedAndPaged has an uncommented print statement

File: src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import java.util.logging.Logger;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java]: Copyright header not updated

File: src/main/src/main/java/org/geoserver/security/decorators/DecoratingCatalogFactory.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
+package org.geoserver.security.decorators;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/DecoratingCatalogFactory.java]: No copyright header

File: src/main/src/main/java/org/geoserver/security/decorators/DecoratingCatalogFactory.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import org.geoserver.catalog.MapInfo;
+import org.geoserver.catalog.MetadataLinkInfo;
+import org.geoserver.catalog.NamespaceInfo;
+import org.geoserver.catalog.StyleInfo;
+import org.geoserver.catalog.WMSLayerInfo;
+import org.geoserver.catalog.WMSStoreInfo;
+import org.geoserver.catalog.WorkspaceInfo;
+import org.geoserver.catalog.impl.AbstractDecorator;
+
+public class DecoratingCatalogFactory extends AbstractDecorator<CatalogFactory> implements CatalogFactory {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/DecoratingCatalogFactory.java]: No javadoc for a public class?

File: src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java [hunk 6]
```diff
@@ -194,7 +195,13 @@ public int compare(Class c1, Class c2) {
             // nothing to copy
             return null;
         }
-        Collection<T> copy = source.getClass().newInstance();
+        Collection<T> copy;
+        try {
+            copy = source.getClass().newInstance();
+        } catch (InstantiationException e) {
+            //we'll just pick something
+            copy = new ArrayList<T>();
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java]: This something is I suppose what you needed, but what if the original collection was a Set or a Map?
What was the class that could not be instantiated?

**Comment by aaime** [comment]
I commented on a few, but several classes are missing copyright year updates, please fix them all.

File: src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java (hunk 6 above)
Comment (NielsCharlier) [inline src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java]: Thanks for the review. The problem is [this](https://github.com/NielsCharlier/geoserver/commit/e919df627299543e2177fc8f814f8a3dfce92f5b#diff-56150f92aa7ba936e390c7a4aa0438ffR50) inner class I made ( a list). I searched for a way to reliably clone any collection easily, but that doesn't seem to exist except if you use some third part library. But I guess the least I could do is add some extra ifs for sets etc... although that shouldn't occur (for now)

Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxyCloner.java]: You could switch the class to static and pass around the fields it needs, then give it also a empty constructor, that would fix issues without assuming a non cloneable collection is necessarily a list.

**Comment by NielsCharlier** [comment]
Fixed

**Pull request merged**